lib/ddtrace/contrib/rails/action_view.rb

```diff
@@ -10,96 +10,86 @@ def self.instrument
 
           # subscribe when the template rendering starts
-          ::ActiveSupport::Notifications.subscribe('start_render_template.action_view') do |*args|
+          ::ActiveSupport::Notifications.subscribe('!datadog.start_render_template.action_view') do |*args|
             start_render_template(*args)
           end
 
-          # subscribe when the partial rendering starts
-          ::ActiveSupport::Notifications.subscribe('start_render_partial.action_view') do |*args|
-            start_render_partial(*args)
+          # subscribe when the template rendering has been processed
+          ::ActiveSupport::Notifications.subscribe('!datadog.finish_render_template.action_view') do |*args|
+            finish_render_template(*args)
           end
 
-          # subscribe when the template rendering has been processed
-          ::ActiveSupport::Notifications.subscribe('render_template.action_view') do |*args|
-            render_template(*args)
+          # subscribe when the partial rendering starts
+          ::ActiveSupport::Notifications.subscribe('!datadog.start_render_partial.action_view') do |*args|
+            start_render_partial(*args)
           end
 
           # subscribe when the partial rendering has been processed
-          ::ActiveSupport::Notifications.subscribe('render_partial.action_view') do |*args|
-            render_partial(*args)
+          ::ActiveSupport::Notifications.subscribe('!datadog.finish_render_partial.action_view') do |*args|
+            finish_render_partial(*args)
           end
         end
 
-        def self.get_key(f)
-          'datadog_actionview_' + f
-        end
-
-        def self.start_render_template(*)
-          key = get_key('render_template')
-          return if Thread.current[key]
+        def self.start_render_template(_name, _start, _finish, _id, payload)
+          # retrieve the tracing context
+          tracing_context = payload.fetch(:tracing_context)
 
+          # create a new Span and add it to the tracing context
           tracer = ::Rails.configuration.datadog_trace.fetch(:tracer)
-          type = Datadog::Ext::HTTP::TEMPLATE
-          tracer.trace('rails.render_template', span_type: type)
-
-          Thread.current[key] = true
+          span = tracer.trace('rails.render_template', span_type: Datadog::Ext::HTTP::TEMPLATE)
+          tracing_context[:dd_rails_template_span] = span
         rescue StandardError => e
-          Datadog::Tracer.log.error(e.message)
+          Datadog::Tracer.log.debug(e.message)
         end
 
-        def self.start_render_partial(*)
-          key = get_key('render_partial')
-          return if Thread.current[key]
+        def self.finish_render_template(_name, _start, _finish, _id, payload)
+          # retrieve the tracing context and the latest active span
+          tracing_context = payload.fetch(:tracing_context)
+          span = tracing_context[:dd_rails_template_span]
+          return if span.try(:finished?)
```

Comment:
I just realized I misunderstood the previous version of this. My fault!
We should also return if `span` is `nil`, right?
So I guess it should be:
```rb
return if !span || span.finished?
```
Sorry for that 😕

Reply:
yep!